processor/samplingprocessor/tailsamplingprocessor/forwarder.go

```diff
@@ -0,0 +1,382 @@
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"net"
+	"net/http"
+	"reflect"
+	"runtime"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.uber.org/zap"
+
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	v1 "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/open-telemetry/opentelemetry-collector/config/configgrpc"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/exporter"
+	"github.com/open-telemetry/opentelemetry-collector/exporter/opencensusexporter"
+	"github.com/open-telemetry/opentelemetry-collector/processor/samplingprocessor/tailsamplingprocessor/idbatcher"
+)
+
+type forwarder interface {
+	// process span
+	process(span *tracepb.Span) bool
+}
+
+type collectorPeer struct {
+	exporter           exporter.TraceExporter
+	peerBatcher        idbatcher.Batcher
+	spanDispatchTicker tTicker // to pop from the batcher and forward to peer
+	logger             *zap.Logger
+	start              sync.Once
+	idToSpans          sync.Map
+	globalDeleteChan   chan int
+}
+
+type ringMembershipExtensionClient struct {
+	client   *http.Client
+	endpoint string
+}
+
+// spanForwarder is the component that fowards spans to collector peers
+// based on traceID
+type spanForwarder struct {
+	// to make peerQueues concurrently accessible
+	sync.RWMutex
+	// to start the memberSyncTicker
+	start sync.Once
+	// logger
+	logger *zap.Logger
+	// self member IP
+	selfMemberIP string
+	// ticker to call ring.GetState() and sync member list
+	memberSyncTicker tTicker
+	// stores queues for each of the collector peers
+	peerQueues map[string]*collectorPeer
+	// The ringmembership extension (implements extension.SupportExtension)
+	// Ownership should be with the component that requires it.
+	// However it will be Start()'ed by the Application service.
+	// ringMembershipExtensionClient is an http client to get extension state
+	ring *ringMembershipExtensionClient
+	// channel to keep track of total number of traces in memory
+	// across all queues
+	// TODO(@annanay25): Is it good to use channels for this
+	globalDeleteChan chan int
+}
+
+func (c *collectorPeer) batchDispatchOnTick() {
+	c.logger.Debug("Collector peer batchDispatchOnTick invoked")
+	batchIds, _ := c.peerBatcher.CloseCurrentAndTakeFirstBatch()
+	// create batch from batchIds
+	var td consumerdata.TraceData
+	for _, v := range batchIds {
+		if span, ok := c.idToSpans.Load(string(v)); ok {
+			if span != nil {
+				span.(*v1.Span).Attributes.AttributeMap["otelcol.ttl"] = &v1.AttributeValue{
+					Value: &v1.AttributeValue_IntValue{
+						IntValue: 1,
+					},
+				}
+				td.Spans = append(td.Spans, span.(*v1.Span))
+				c.logger.Info("Forwarding this span", zap.ByteString("Span ID", span.(*v1.Span).GetSpanId()))
```

Comment:
This will flood the log, we don't want to use Info level logging for spans. See https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#logging